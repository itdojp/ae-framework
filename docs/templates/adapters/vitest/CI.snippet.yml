name: adapter-vitest
on: [push]
jobs:
  vitest:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with: { node-version: '20' }
      - run: npm ci
      - run: npx vitest run --reporter=dot
      - name: Write normalized summary
        run: |
          mkdir -p artifacts/vitest
          cat > artifacts/vitest/summary.json << 'JSON'
          {"adapter":"vitest","status":"ok","summary":"Vitest: 220 passed, 0 failed","details":[]}
          JSON
      - name: Upload artifact
        uses: actions/upload-artifact@v4
        with:
          name: artifacts-vitest
          path: artifacts/vitest/summary.json
