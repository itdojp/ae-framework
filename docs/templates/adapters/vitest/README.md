# Vitest Adapter Templates (Examples)

Vitest can emit JSON summary which can be aggregated in PR comments.

- Write a minimal JSON summary to `artifacts/adapters/vitest/summary.json`.
- Upload the directory as artifact in CI, and aggregate via `scripts/adapters/aggregate-artifacts.mjs`.

Minimal CI snippet:
```yaml
- name: Aggregate Vitest artifacts
  run: node scripts/adapters/aggregate-artifacts.mjs || true
  if: always()
```

Minimal JSON shape (`artifacts/adapters/vitest/summary.json`):
```json
{
  "adapter": "vitest",
  "summary": "passed=200 failed=0",
  "status": "ok",
  "traceId": null
}
```

Alternative compact shape:
```
{
  "tool": "vitest",
  "summary": { "passed": 42, "failed": 0, "skipped": 0 },
  "suites": [ { "name": "unit", "passed": 30, "failed": 0 } ]
}
```
Aggregated by `scripts/adapters/aggregate-artifacts.mjs` into a single PR summary.
