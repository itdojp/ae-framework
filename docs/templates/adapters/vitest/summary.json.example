{
  "adapter": "vitest",
  "status": "ok",
  "summary": "Vitest: 220 passed, 0 failed",
  "details": [
    { "file": "aggregate.invariants.test.ts", "passed": 20, "failed": 0 }
  ]
}
