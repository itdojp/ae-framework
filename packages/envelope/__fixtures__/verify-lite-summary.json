{
  "schemaVersion": "1.0.0",
  "timestamp": "2025-10-06T02:29:59.000Z",
  "flags": {
    "install": "--frozen-lockfile",
    "noFrozen": false,
    "keepLintLog": true,
    "enforceLint": true,
    "runMutation": true
  },
  "steps": {
    "install": { "status": "success", "retried": false },
    "lint": { "status": "failure", "notes": "baseline +5" },
    "mutationQuick": { "status": "success" }
  },
  "artifacts": {
    "lintSummary": "verify-lite-lint-summary.json",
    "lintLog": null,
    "mutationSummary": "reports/mutation/summary.json",
    "mutationSurvivors": "reports/mutation/survivors.json",
    "conformanceSummary": "reports/conformance/verify-lite-summary.json",
    "conformanceSummaryMarkdown": null
  },
  "trace": {
    "status": "valid",
    "traceIds": ["trace-1"],
    "tempoLinks": ["https://tempo.example.com/explore?traceId=trace-1"],
    "domains": [
      {
        "key": "inventory",
        "traceIds": ["inventory-trace"],
        "tempoLinks": ["https://tempo.example.com/explore?traceId=inventory-trace"]
      }
    ]
  },
  "tempoLinks": ["https://tempo.example.com/explore?traceId=summary-trace"]
}
