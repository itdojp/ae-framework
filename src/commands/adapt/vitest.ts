import * as fs from 'fs';
import * as path from 'path';
import chalk from 'chalk';

interface PackageJson {
  scripts?: Record<string, string>;
  [key: string]: any;
}

function generateVitestConfigTemplate(thresholds = { lines: 80, functions: 80, branches: 80, statements: 80 }) {
  return `import { defineConfig } from 'vitest/config';

export default defineConfig({
  test: {
    coverage: {
      provider: 'v8', // or 'c8'
      thresholds: {
        lines: ${thresholds.lines},
        functions: ${thresholds.functions},
        branches: ${thresholds.branches},
        statements: ${thresholds.statements},
      }
    }
  }
});
`;
}

function backupFile(filePath: string): void {
  if (fs.existsSync(filePath)) {
    const backupPath = `${filePath}.bak`;
    fs.copyFileSync(filePath, backupPath);
    console.log(chalk.blue(`üìã Backed up ${path.basename(filePath)} to ${path.basename(backupPath)}`));
  }
}

function updatePackageJson(): boolean {
  const packageJsonPath = path.join(process.cwd(), 'package.json');
  
  if (!fs.existsSync(packageJsonPath)) {
    console.log(chalk.red('‚ùå package.json not found'));
    return false;
  }

  backupFile(packageJsonPath);
  
  const packageJson: PackageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
  
  let modified = false;

  // Add test script if missing
  if (!packageJson.scripts) {
    packageJson.scripts = {};
  }

  if (!packageJson.scripts.test) {
    const seedCmd = process.env.AE_SEED ? `AE_SEED=${process.env.AE_SEED} ` : '';
    packageJson.scripts.test = `${seedCmd}vitest run --coverage`;
    modified = true;
    console.log(chalk.green('‚úÖ Added test script with coverage'));
  } else if (!packageJson.scripts.test.includes('--coverage') && packageJson.scripts.test.includes('vitest')) {
    // Update existing vitest script to include coverage
    const currentTest = packageJson.scripts.test;
    packageJson.scripts.test = `${currentTest} --coverage`;
    modified = true;
    console.log(chalk.green('‚úÖ Updated test script with coverage'));
  } else if (!packageJson.scripts.test.includes('vitest')) {
    console.log(chalk.yellow('‚ö†Ô∏è  Existing test script does not use Vitest, skipping coverage injection'));
  } else {
    console.log(chalk.blue('‚ÑπÔ∏è  Test script already has coverage configuration'));
  }

  if (modified) {
    fs.writeFileSync(packageJsonPath, JSON.stringify(packageJson, null, 2) + '\n');
  }

  return true;
}

function generateVitestConfigSnippet(thresholds = { lines: 80, functions: 80, branches: 80, statements: 80 }) {
  return `
  test: {
    coverage: {
      provider: 'v8',
      thresholds: {
        lines: ${thresholds.lines},
        functions: ${thresholds.functions}, 
        branches: ${thresholds.branches},
        statements: ${thresholds.statements},
      }
    }
  }`;
}

function createVitestConfig(customThresholds?: { statements: number; branches: number; functions: number; lines: number }): void {
  const configPaths = [
    'vitest.config.ts',
    'vitest.config.js',
    'vite.config.ts',
    'vite.config.js'
  ];
  
  const existingConfig = configPaths.find(p => fs.existsSync(path.join(process.cwd(), p)));
  
  const thresholds = customThresholds || { lines: 80, functions: 80, branches: 80, statements: 80 };

  if (existingConfig) {
    console.log(chalk.blue(`‚ÑπÔ∏è  Existing config found: ${existingConfig}`));
    console.log(chalk.yellow('‚ö†Ô∏è  Please manually add coverage thresholds to your existing config:'));
    console.log(chalk.cyan(generateVitestConfigSnippet(thresholds)));
    return;
  }

  const configPath = path.join(process.cwd(), 'vitest.config.ts');
  fs.writeFileSync(configPath, generateVitestConfigTemplate(thresholds));
  console.log(chalk.green('‚úÖ Created vitest.config.ts with coverage thresholds'));
}

function updatePreCommitHook(): void {
  const huskyPath = path.join(process.cwd(), '.husky');
  const preCommitPath = path.join(huskyPath, 'pre-commit');
  
  if (!fs.existsSync(preCommitPath)) {
    console.log(chalk.blue('‚ÑπÔ∏è  No .husky/pre-commit found, skipping guard setup'));
    return;
  }

  backupFile(preCommitPath);
  
  const preCommitContent = fs.readFileSync(preCommitPath, 'utf8');
  
  if (preCommitContent.includes('ae tdd:guard')) {
    console.log(chalk.blue('‚ÑπÔ∏è  TDD guard already configured in pre-commit hook'));
    return;
  }

  const updatedContent = preCommitContent.trim() + '\nae tdd:guard\n';
  fs.writeFileSync(preCommitPath, updatedContent);
  console.log(chalk.green('‚úÖ Added TDD guard to pre-commit hook'));
}

export async function adaptVitest(thresholds?: { statements: number; branches: number; functions: number; lines: number }) {
  console.log(chalk.blue('üîß Adapting project for Vitest with ae-framework integration\n'));

  try {
    const success = updatePackageJson();
    if (!success) return;

    createVitestConfig(thresholds);
    updatePreCommitHook();

    console.log(chalk.cyan('\nüìã Vitest Adaptation Summary:'));
    console.log(chalk.cyan('  ‚Ä¢ Coverage thresholds: 80% for all metrics'));
    console.log(chalk.cyan('  ‚Ä¢ AE_SEED integration for deterministic tests'));
    console.log(chalk.cyan('  ‚Ä¢ TDD guard integration via pre-commit hook'));
    console.log(chalk.cyan('  ‚Ä¢ Coverage provider: v8 (configurable)'));
    
    console.log(chalk.cyan('\nüîÑ Next Steps:'));
    console.log(chalk.cyan('  ‚Ä¢ Install coverage provider: npm i -D @vitest/coverage-v8'));
    console.log(chalk.cyan('  ‚Ä¢ Run: npm test'));
    console.log(chalk.cyan('  ‚Ä¢ Verify coverage thresholds are enforced'));
    console.log(chalk.cyan('  ‚Ä¢ Check pre-commit hook with: git commit'));
    
    console.log(chalk.green('\n‚úÖ Vitest adaptation completed successfully!'));

  } catch (error) {
    console.error(chalk.red(`‚ùå Vitest adaptation failed: ${error instanceof Error ? error.message : 'Unknown error'}`));
    process.exit(1);
  }
}