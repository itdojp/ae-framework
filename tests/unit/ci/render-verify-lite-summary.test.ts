import { describe, expect, it } from 'vitest';
import { renderVerifyLiteSummary } from '../../../scripts/ci/lib/verify-lite-summary.mjs';

describe('renderVerifyLiteSummary', () => {
  const baseSummary = {
    schemaVersion: '1.0.0',
    timestamp: '2025-10-06T00:00:00Z',
    flags: {
      install: '--frozen-lockfile',
      noFrozen: false,
      keepLintLog: true,
      enforceLint: false,
      runMutation: true
    },
    steps: {
      install: { status: 'success', notes: 'flags=--frozen-lockfile', retried: false },
      lint: { status: 'failure', notes: '2618 violations' },
      build: { status: 'success' },
      bddLint: { status: 'skipped' },
      mutationQuick: { status: 'success', notes: 'score: 59.74%' },
      conformanceReport: { status: 'success', notes: 'runs=1;violations=0' }
    },
    artifacts: {
      lintSummary: 'verify-lite-lint-summary.json',
      lintLog: 'verify-lite-lint.log',
      mutationSummary: 'reports/mutation/summary.json',
      mutationSurvivors: 'reports/mutation/survivors.json',
      conformanceSummary: 'reports/conformance/verify-lite-summary.json',
      conformanceSummaryMarkdown: 'reports/conformance/verify-lite-summary.md'
    }
  };

  it('renders markdown summary with schema version and flags', () => {
    const result = renderVerifyLiteSummary(baseSummary, { artifactsUrl: 'https://example.com/artifacts' });
    expect(result).toMatchSnapshot();
  });

  it('handles missing optional notes and artifacts', () => {
    const minimal = {
      schemaVersion: '1.0.0',
      timestamp: '2025-10-06T01:23:45Z',
      flags: {
        install: '',
        noFrozen: true,
        keepLintLog: false,
        enforceLint: false,
        runMutation: false
      },
      steps: {
        typeCheck: { status: 'success' },
        lint: { status: 'skipped' }
      },
      artifacts: {
        lintSummary: null,
        lintLog: null,
        mutationSummary: null,
        mutationSurvivors: null,
        conformanceSummary: null,
        conformanceSummaryMarkdown: null
      }
    };

    const result = renderVerifyLiteSummary(minimal);
    expect(result).toMatchSnapshot();
  });

  it('escapes HTML-sensitive characters in notes', () => {
    const summary = {
      ...baseSummary,
      steps: {
        ...baseSummary.steps,
        lint: { status: 'failure', notes: '<script>alert("xss")</script>' }
      }
    };

    const result = renderVerifyLiteSummary(summary);
    expect(result).toContain('&lt;script&gt;alert(&quot;xss&quot;)&lt;/script&gt;');
  });

  it('throws on invalid payload', () => {
    // @ts-expect-error deliberate bad input
    expect(() => renderVerifyLiteSummary(null)).toThrowErrorMatchingInlineSnapshot(
      "[Error: Invalid summary payload]"
    );
  });
});
