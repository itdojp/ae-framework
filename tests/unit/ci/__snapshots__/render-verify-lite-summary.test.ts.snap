// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderVerifyLiteSummary > handles missing optional notes and artifacts 1`] = `
"Timestamp: 2025-10-06T01:23:45Z
Schema Version: 1.0.0
- install flags: \`\`
- no frozen lockfile: ✅
- keep lint log: ❌
- enforce lint: ❌
- run mutation: ❌

| Step | Status | Notes |
| --- | --- | --- |
| Type Check | ✅ success |  |
| Lint | ⏭️ skipped |  |


Artifacts:
- lintSummary: n/a
- lintLog: n/a
- mutationSummary: n/a
- mutationSurvivors: n/a"
`;

exports[`renderVerifyLiteSummary > renders markdown summary with schema version and flags 1`] = `
"Timestamp: 2025-10-06T00:00:00Z
Schema Version: 1.0.0
- install flags: \`--frozen-lockfile\`
- no frozen lockfile: ❌
- keep lint log: ✅
- enforce lint: ❌
- run mutation: ✅

| Step | Status | Notes |
| --- | --- | --- |
| Install | ✅ success | flags=--frozen-lockfile |
| Lint | ❌ failure | 2618 violations |
| Build | ✅ success |  |
| Bdd Lint | ⏭️ skipped |  |
| Mutation Quick | ✅ success | score: 59.74% |


Artifacts:
- lintSummary: verify-lite-lint-summary.json
- lintLog: verify-lite-lint.log
- mutationSummary: mutation-summary.md
- mutationSurvivors: reports/mutation/survivors.json"
`;
